browser/ntp_sponsored_images/new_tab_page_branded_view_counter.cc

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2020 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <memory>
+#include <string>
+
+#include "base/bind.h"
+#include "base/no_destructor.h"
+#include "base/strings/stringprintf.h"
+#include "brave/browser/brave_browser_process_impl.h"
+#include "brave/browser/ntp_sponsored_images/new_tab_page_branded_view_counter.h"
+#include "brave/common/pref_names.h"
+#include "brave/components/brave_ads/browser/ads_service.h"
+#include "brave/components/brave_ads/browser/ads_service_factory.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
+#include "brave/components/brave_rewards/browser/rewards_service_factory.h"
+#include "chrome/browser/profiles/incognito_helpers.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/common/chrome_features.h"
+#include "content/public/browser/browser_context.h"
+#include "components/keyed_service/content/browser_context_dependency_manager.h"
+#include "components/keyed_service/content/browser_context_keyed_service_factory.h"
+#include "components/prefs/pref_change_registrar.h"
+#include "components/prefs/pref_service.h"
+
+namespace {
+
+std::unique_ptr<NTPSponsoredImagesData> GetDemoWallpaper() {
+  auto demo = std::make_unique<NTPSponsoredImagesData>();
+  demo->wallpaper_image_urls = {
+      "ntp-dummy-brandedwallpaper-background-1.jpg",
+      "ntp-dummy-brandedwallpaper-background-2.jpg",
+      "ntp-dummy-brandedwallpaper-background-3.jpg"};
+  demo->logo_image_url = "ntp-dummy-brandedwallpaper-logo.png";
+  demo->logo_alt_text = "Technikke: For music lovers.";
+  demo->logo_company_name = "Technikke";
+  demo->logo_destination_url = "https://brave.com";
+  return demo;
+}
+
+class NewTabPageBrandedViewCounterFactory
+    : public BrowserContextKeyedServiceFactory {
+ public:
+  static NewTabPageBrandedViewCounter* GetForProfile(Profile* profile) {
+    return static_cast<NewTabPageBrandedViewCounter*>(
+        GetInstance()->GetServiceForBrowserContext(profile, true));
+  }
+
+  static NewTabPageBrandedViewCounterFactory* GetInstance() {
+    static base::NoDestructor<NewTabPageBrandedViewCounterFactory> instance;
+    return instance.get();
+  }
+  NewTabPageBrandedViewCounterFactory() : BrowserContextKeyedServiceFactory(
+        "NewTabPageBrandedViewCounter",
+        BrowserContextDependencyManager::GetInstance()) {}
```

Comment:
this needs `DependsOn` for brave rewards and brave ads to ensure they are started first

Reply:
otherwise https://github.com/brave/brave-core/pull/4075/files#diff-3fa0819ac2021e1ea91ba79a7d6adfa6R110 will be a problem

Reply:
done